### PR TITLE
Bug 1270728 - username in the secret don't override the username in the source URL

### PR DESCRIPTION
**Pull request by jhadvig**
Returning the url in the `password.go` and `cacert.go`
@csrwng PTAL

### CODE REVIEW

File: pkg/build/builder/cmd/builder.go [hunk 1]
```diff
@@ -106,19 +110,21 @@ func setupSourceSecret(sourceSecretName string, scmAuths []scmauth.SCMAuth) erro
 	}
 
 	if len(scmAuthsPresent) == 0 {
-		return fmt.Errorf("no auth handler was found for the provided secret %q",
+		return nil, fmt.Errorf("no auth handler was found for the provided secret %q",
 			sourceSecretName)
 	}
 
+	url := &url.URL{}
```
Comment (mfojtik) [inline pkg/build/builder/cmd/builder.go]: This is wrong, you're shadowing the package name :-) Use different variable name.

File: pkg/build/builder/cmd/scmauth/scmauth.go [hunk 2]
```diff
@@ -11,5 +13,5 @@ type SCMAuth interface {
 	Handles(name string) bool
 
 	// Setup lays down the required files for this authentication method to work
```
Comment (liggitt) [inline pkg/build/builder/cmd/scmauth/scmauth.go]: document what the returned URL is... if the auth methods mutate the URL now, their order becomes important... how is that determined? are there test cases around it?

File: pkg/build/builder/cmd/scmauth/password.go [hunk 3]
```diff
@@ -86,7 +86,7 @@ func (u UsernamePassword) Setup(baseDir string) error {
 	fmt.Fprintf(gitconfig, UserPassGitConfig, gitcredentials.Name())
 	fmt.Fprintf(gitcredentials, "%s", usernamePasswordURL.String())
 
-	return ensureGitConfigIncludes(gitconfig.Name())
+	return &usernamePasswordURL, ensureGitConfigIncludes(gitconfig.Name())
```
Comment (csrwng) [inline pkg/build/builder/cmd/scmauth/password.go]: Actually the URL returned should be the URL without Username/Pass. That way the credentials are not output in the logs.
git will then use the credential store and the gitcredentials file to figure out the credentials to use.

File: pkg/build/builder/cmd/builder.go (hunk 1 above)
Comment (csrwng) [inline pkg/build/builder/cmd/builder.go]: You can also just declare the variable without allocating a value for it (var u *url.URL)

File: pkg/build/builder/cmd/builder.go [hunk 4]
```diff
@@ -106,19 +110,21 @@ func setupSourceSecret(sourceSecretName string, scmAuths []scmauth.SCMAuth) erro
...
+
 	for name, auth := range scmAuthsPresent {
 		glog.V(3).Infof("Setting up SCMAuth %q", name)
-		if err := auth.Setup(sourceSecretDir); err != nil {
+		if url, err = auth.Setup(sourceSecretDir); err != nil {
```
Comment (liggitt) [inline pkg/build/builder/cmd/builder.go]: this is hard to follow... the last auth will win, even if an earlier one set to non-nil

File: pkg/build/builder/cmd/builder.go [hunk 5]
```diff
@@ -106,19 +110,21 @@ func setupSourceSecret(sourceSecretName string, scmAuths []scmauth.SCMAuth) erro
...
 			// If an error occurs during setup, fail the build
-			return fmt.Errorf("cannot set up source authentication method %q: %v", name, err)
+			return nil, fmt.Errorf("cannot set up source authentication method %q: %v", name, err)
 		}
 	}
```
Comment (liggitt) [inline pkg/build/builder/cmd/builder.go]: I think there should be a second step here where the auths are asked for an override URL... all the non-nil ones are collected. If two non-nil URLs differ, that's probably an error

File: pkg/build/builder/cmd/builder.go (hunk 4 above)
Comment (csrwng) [inline pkg/build/builder/cmd/builder.go]: actually I would use a temp variable here and only if it's non-nil assign it to the outer variable. Otherwise if you have multiple scmAuth's, one that returns a nil value will overwrite the previous scmAuth's URL. I would also either log a warning or throw an error if multiple scmAuth's return non-nil for the URL.

File: pkg/build/builder/cmd/builder.go [hunk 6]
```diff
@@ -45,9 +45,13 @@ func run(builderFactory factoryFunc) {
 			glog.Fatalf("Cannot parse build URL: %s", build.Spec.Source.Git.URI)
 		}
 		scmAuths := auths(sourceURL)
-		if err := setupSourceSecret(build.Spec.Source.SourceSecret.Name, scmAuths); err != nil {
+		url, err := setupSourceSecret(build.Spec.Source.SourceSecret.Name, scmAuths)
+		if err != nil {
 			glog.Fatalf("Cannot setup secret file for accessing private repository: %v", err)
 		}
+		if url != nil {
+			build.Spec.Source.Git.URI = url.String()
```
Comment (liggitt) [inline pkg/build/builder/cmd/builder.go]: is stomping the entire URL correct? we need good tests to make sure all the non-auth-related bits of the URL are preserved (path, etc)

File: pkg/build/builder/cmd/builder.go [hunk 7]
```diff
@@ -45,9 +45,13 @@ func run(builderFactory factoryFunc) {
 			glog.Fatalf("Cannot parse build URL: %s", build.Spec.Source.Git.URI)
 		}
 		scmAuths := auths(sourceURL)
-		if err := setupSourceSecret(build.Spec.Source.SourceSecret.Name, scmAuths); err != nil {
+		url, err := setupSourceSecret(build.Spec.Source.SourceSecret.Name, scmAuths)
```
Comment (liggitt) [inline pkg/build/builder/cmd/builder.go]: alternately, pass in the `build.Spec.Source.Git.URI` and assign the output, so each auth gets a crack at mutating it as appropriate

**Comment by jhadvig** [comment]
@csrwng @liggitt comments addressed. PTAL

File: pkg/build/builder/cmd/builder.go [hunk 8]
```diff
@@ -45,9 +45,13 @@ func run(builderFactory factoryFunc) {
 				glog.Fatalf("Cannot parse build URL: %s", build.Spec.Source.Git.URI)
 			}
 			scmAuths := auths(sourceURL)
-			if err := setupSourceSecret(build.Spec.Source.SourceSecret.Name, scmAuths); err != nil {
+			sourceURL, err = setupSourceSecret(build.Spec.Source.SourceSecret.Name, scmAuths)
```
Comment (liggitt) [inline pkg/build/builder/cmd/builder.go]: don't you have to pass the original URI in?

Comment (csrwng) [inline pkg/build/builder/cmd/builder.go]: It's passed in when the scm auth is created: https://github.com/jhadvig/origin/blob/BZ_1270728/pkg/build/builder/cmd/builder.go#L146-L147

File: pkg/build/builder/cmd/scmauth/password.go [hunk 9]
```diff
@@ -29,19 +30,19 @@ type UsernamePassword struct {
...
-		return nil
+		return nil, nil
 	}
 
 	// Determine username
 	// 1. Look for a username secret
 	username, err := readSecret(baseDir, UsernameSecret)
 	if err != nil {
-		return err
+		return nil, nil
```
Comment (liggitt) [inline pkg/build/builder/cmd/scmauth/password.go]: why are we no longer returning errors here?

File: pkg/build/builder/cmd/scmauth/password.go [hunk 10]
```diff
@@ -56,13 +57,13 @@ func (u UsernamePassword) Setup(baseDir string) error {
 	// 1. Look for a token secret
 	password, err := readSecret(baseDir, TokenSecret)
 	if err != nil {
-		return err
+		return nil, nil
```
Comment (liggitt) [inline pkg/build/builder/cmd/scmauth/password.go]: why are we no longer returning errors here?

File: pkg/build/builder/cmd/scmauth/password.go [hunk 11]
```diff
@@ -56,13 +57,13 @@ func (u UsernamePassword) Setup(baseDir string) error {
...
 	}
 	// 2. Look for a password secret
 	if password == "" {
 		password, err = readSecret(baseDir, PasswordSecret)
 		if err != nil {
-			return err
+			return nil, nil
```
Comment (liggitt) [inline pkg/build/builder/cmd/scmauth/password.go]: why are we no longer returning errors here?

File: pkg/build/builder/cmd/scmauth/password.go [hunk 12]
```diff
@@ -72,12 +73,12 @@ func (u UsernamePassword) Setup(baseDir string) error {
 
 	gitcredentials, err := ioutil.TempFile("", "gitcredentials.")
 	if err != nil {
-		return err
+		return nil, nil
```
Comment (liggitt) [inline pkg/build/builder/cmd/scmauth/password.go]: why are we no longer returning errors here?

File: pkg/build/builder/cmd/scmauth/password.go [hunk 13]
```diff
@@ -86,7 +87,12 @@ func (u UsernamePassword) Setup(baseDir string) error {
 	fmt.Fprintf(gitconfig, UserPassGitConfig, gitcredentials.Name())
 	fmt.Fprintf(gitcredentials, "%s", usernamePasswordURL.String())
 
-	return ensureGitConfigIncludes(gitconfig.Name())
+	sourceURL := u.SourceURL
+	sourceURL.User = nil
```
Comment (liggitt) [inline pkg/build/builder/cmd/scmauth/password.go]: don't we have to leave the username, and strip the password? From the git docs: "If the protocol, hostname, and username (if we already have one) match..."

Comment (liggitt) [inline pkg/build/builder/cmd/scmauth/password.go]: I would expect `sourceURL.User = username`

File: pkg/build/builder/cmd/scmauth/sshkey.go [hunk 14]
```diff
@@ -13,25 +14,25 @@ type SSHPrivateKey struct{}
...
-		return err
+		return nil, err
 	}
 	// set environment variable to tell git to use the SSH wrapper
 	if err := os.Setenv("GIT_SSH", script.Name()); err != nil {
-		return err
+		return nil, err
 	}
-	return nil
+	return nil, nil
```
Comment (liggitt) [inline pkg/build/builder/cmd/scmauth/sshkey.go]: when using ssh, should we remove passwords from the URL if present?

File: pkg/build/builder/cmd/scmauth/password.go (hunk 13 above)
Comment (csrwng) [inline pkg/build/builder/cmd/scmauth/password.go]: Both username and password would come from the credentials file. They're not needed in the URL

File: pkg/build/builder/cmd/scmauth/password.go [hunk 15]
```diff
@@ -86,7 +87,12 @@ func (u UsernamePassword) Setup(baseDir string) error {
...
+
+	glog.V(1).Infof("-------> %q", sourceURL.String())
```
Comment (liggitt) [inline pkg/build/builder/cmd/scmauth/password.go]: remove debug output?

File: pkg/build/builder/cmd/scmauth/sshkey.go (hunk 14 above)
Comment (csrwng) [inline pkg/build/builder/cmd/scmauth/sshkey.go]: can you specify a password in an ssh spec?

File: pkg/build/builder/cmd/builder.go [hunk 16]
```diff
@@ -107,19 +111,33 @@ func setupSourceSecret(sourceSecretName string, scmAuths []scmauth.SCMAuth) erro
...
 	}
 
-	return nil
+	if len(urlArray) == 0 {
+		return nil, nil
+	}
+
+	for i := range urlArray {
+		if urlArray[i].String() != urlArray[0].String() {
+			return nil, fmt.Errorf("secrets contain different source URLs")
```
Comment (liggitt) [inline pkg/build/builder/cmd/builder.go]: would be helpful to give more info... can we mask the passwords in the urls and include them in the message?

File: pkg/build/builder/cmd/scmauth/password.go (hunk 13 above)
Comment (liggitt) [inline pkg/build/builder/cmd/scmauth/password.go]: username doesn't have to match from the URL? I couldn't tell from the documentation

File: pkg/build/builder/cmd/scmauth/sshkey.go (hunk 14 above)
Comment (liggitt) [inline pkg/build/builder/cmd/scmauth/sshkey.go]: I don't know, that's why it was a question :)

Comment (csrwng) [inline pkg/build/builder/cmd/scmauth/sshkey.go]: From the git doc, it doesn't look like you can... git will only pick up the ssh private key if using a url like user@host:/path which shouldn't contain a password

File: pkg/build/builder/cmd/scmauth/password.go (hunk 13 above)
Comment (csrwng) [inline pkg/build/builder/cmd/scmauth/password.go]: From here https://git-scm.com/docs/git-credential-store (at the end) it looks like git will try to match the username if one is provided in the URL. However if we strip the username and password from the URL, it should match on protocol and hostname which is fine, since we're only putting a single entry there. Because it will show in the logs, I'd prefer leaving the name out especially if you consider it a secret.

Comment (liggitt) [inline pkg/build/builder/cmd/scmauth/password.go]: I guess that's ok... just trying to anticipate advanced use cases where the user provided a gitconfig of their own, and might be counting on the username match to select the right credential

**Comment by jhadvig** [comment]
@liggitt thanks for the comments. I've updated the PR. Left the the comments as they are since I see @csrwng explained the use-cases.

File: pkg/build/builder/cmd/builder.go [hunk 17]
```diff
@@ -107,19 +111,33 @@ func setupSourceSecret(sourceSecretName string, scmAuths []scmauth.SCMAuth) erro
...
 	}
 
-	return nil
+	if len(urlArray) == 0 {
+		return nil, nil
+	}
+
+	for i := range urlArray {
+		if urlArray[i].String() != urlArray[0].String() {
+			return nil, fmt.Errorf("secret handlers contain different source URLs defined for the secret %q", sourceSecretName)
```
Comment (liggitt) [inline pkg/build/builder/cmd/builder.go]: I was actually thinking something like this:
```
var urlOverride *url.URL = nil
...
for name, auth := range scmAuthsPresent {
  ...
  if u != nil {
    if urlOverride == nil {
      urlOverride = u
    } else if urlOverride.String() != u.String() {
      return nil, fmt.Errorf("secret handler %s set a conflicting override URL %q (conflicts with earlier override URL %q)", auth.Name(), u.String(), urlOverride.String())
    }
  }
}

return urlOverride, nil
```

Comment (liggitt) [inline pkg/build/builder/cmd/builder.go]: none of our secret methods return URL overrides containing passwords, which is why I'm ok logging these

File: pkg/build/builder/cmd/scmauth/password.go [hunk 18]
```diff
@@ -72,12 +72,12 @@ func (u UsernamePassword) Setup(baseDir string) error {
```
Comment (liggitt) [inline pkg/build/builder/cmd/scmauth/password.go]: why do we continue if we had no username or password? We could have hit step 3 above and defaulted to `DefaultUsername`, and gotten no password (from either the secret or the URL), and we still continue and write a credential gitconfig. We should only do that if we have password (from the URL or from the secret), or if we have a username (from the secret), right?

File: pkg/build/builder/cmd/scmauth/password.go [hunk 19]
```diff
@@ -86,7 +86,10 @@ func (u UsernamePassword) Setup(baseDir string) error {
 	fmt.Fprintf(gitconfig, UserPassGitConfig, gitcredentials.Name())
 	fmt.Fprintf(gitcredentials, "%s", usernamePasswordURL.String())
 
-	return ensureGitConfigIncludes(gitconfig.Name())
+	sourceURL := u.SourceURL
+	sourceURL.User = nil
+
+	return &sourceURL, ensureGitConfigIncludes(gitconfig.Name())
```
Comment (liggitt) [inline pkg/build/builder/cmd/scmauth/password.go]: I'd like to see the following tests:
* No user/pw in URL, no user/pw/token in secret: no gitconfig written, no URL override returned
* User in URL, no user/pw/token in secret: no gitconfig written, no URL override returned
* User and pw in URL, no user/pw/token in secret: gitconfig written with user/pw from URL, URL override returned with user/pw stripped
* User and pw in URL, user and pw in secret: gitconfig written with user/pw from secret, URL override returned with user/pw stripped
* User in URL only, password in secret only: gitconfig written with user from URL, password from secret, URL override returned with user stripped
* No user/pw in URL, password in secret: gitconfig written with default user, password from secret

File: pkg/build/builder/cmd/scmauth/password.go (hunk 18 above)
Comment (csrwng) [inline pkg/build/builder/cmd/scmauth/password.go]: So I guess we should defer step 3 above for the username, and only do it if we have a password. Otherwise, if no password and no username, return.

Comment (liggitt) [inline pkg/build/builder/cmd/scmauth/password.go]: That's what I would expect. It's weird to get a default username out of the blue... that could even break servers which allow anonymous auth, but not incorrect random usernames

File: pkg/build/builder/cmd/scmauth/password.go (hunk 19 above)
Comment (jhadvig) [inline pkg/build/builder/cmd/scmauth/password.go]: Can we do this as a followup issue, since QA are waiting for the fix ?

Comment (liggitt) [inline pkg/build/builder/cmd/scmauth/password.go]: This is complex enough that we need tests. Opened https://github.com/jhadvig/origin/pull/3 (and https://github.com/openshift/origin/pull/5486 so @csrwng can review) with tests

**Comment by liggitt** [comment]
superceded by #5486
